### DISCUSSION TITLE
Forward Reference Validation

### DISCUSSION THREAD

**Discussion by snarkipus** [eclipse-langium/langium#632] [Q&A] [answered]
Tag: @msujew 
* [Example Repo](https://github.com/snarkipus/langium-expressions)

I'm working through the _Forward Reference_ validation portion in [this book](https://www.amazon.com/Implementing-Domain-Specific-Languages-Xtext-Xtend-ebook-dp-B01CSLI6HM/dp/B01CSLI6HM/ref=mt_other?_encoding=UTF8&me=&qid=) since I'll need that functionality for my implementation. 

For reference, the following are examples of illegal forward references that are otherwise valid.
```js
// Illegal forward reference
var i = j + 1
var j = 0
```
```js
// Illegal mutual dependency
var i = j + 1
var j = i + 1
```
The validator requires a utility class (`ExpressionsModelUtil`) that checks for prior variable declarations before a reference occurs.  As such, I'm trying to build an analog of this class using the Langium framework and TypeScript, hopefully using features in the generated ast.ts file. This author seems to LOVE the dynamic dispatch and extension methods features of Xtend, which I find somewhat non-intuitive as a newcomer. 

Here is the reference Xtext/Xtend implementation.

```xtend
package org.example.expressions

import org.example.expressions.expressions.VariableRef
import org.example.expressions.expressions.Expression
import org.example.expressions.expressions.AbstractElement
import org.example.expressions.expressions.ExpressionsModel
import org.example.expressions.expressions.Variable

import static extension org.eclipse.xtext.EcoreUtil2.*

class ExpressionsModelUtil {
	def isVariableDefinedBefore(VariableRef varRef) {
		varRef.variablesDefinedBefore.contains(varRef.variable)
	}
	
	def variablesDefinedBefore(Expression e) {
		e.getContainerOfType(AbstractElement).variablesDefinedBefore
	}
	
	def variablesDefinedBefore(AbstractElement containingElement) {
		val allElements = (containingElement.eContainer as ExpressionsModel).elements
		
		allElements.subList(0, allElements.indexOf(containingElement)).typeSelect(Variable).toSet
	}
	
}
```
Here is what I've got so far ... which seems to be pretty close ... but I'm somewhat stuck. Before I spend a massive amount of time trying to fix it, I just wanted to make sure this approach was even sound or idiomatic. There's a couple of assumptions regarding Xtext/ECore equivalence I made which may not be valid (i.e. AbstractElement.eContainer() is the same as AbstractElement.$container, getContainerOfType() implementation, etc.).
```ts
import { AstNode } from "langium";
import { VariableRef, Expression, AbstractElement, ExpressionsModel, ExpressionsAstType } from "../language-server/generated/ast";
import { isExpression, isVariable, isAbstractElement, reflection } from "../language-server/generated/ast";

export class ExpressionsModelUtil {

    public isVariableDefinedBefore(varRef: VariableRef): boolean {
        this.variablesDefinedBefore(varRef).has(varRef.variable);
        return true;
    }

    public variablesDefinedBefore(e: Expression | AbstractElement)  {
        // traverse the AST 
        if (isExpression(e)) {
            this.variablesDefinedBefore(getContainerOfType(e, AbstractElement) as Expression | AbstractElement);
        } 
        // collect the variables
        else if (isAbstractElement(e)) {
            const allElements = (e.$container as ExpressionsModel).elements;
            return new Set(allElements
                .slice(0, allElements.indexOf(e))
                .filter(e => isVariable(e)));
        }
    }
}

function getContainerOfType(ele: AstNode, type: ExpressionsAstType ): any {
    for (let e = ele; e !== null; e = e.$container as AstNode) {
        if (reflection.isInstance(e, type)) {
            return e;
        }
        return null;
    }
}
```

### Grammar (partial)
```ts
grammar Expressions

entry ExpressionsModel:
	elements += AbstractElement*
;

AbstractElement:
	Variable | EvalExpression
;

Variable:
	'var' name=ID '=' expression=Expression
;

EvalExpression:
	'eval' expression=Expression
;
...
```

**Comment by msujew** [comment, marked as answer]
Hey @snarkipus,

multiple thoughts on this:

1. We actually implement a `getContainerOfType` function in Langium, which is typed a bit better than yours. You can use it like this:
```ts
const abstractElement = getContainerOfType(e, isAbstractElement); // Through typescript magic, `abstractElement` is automatically typed as `AbstractElement | undefined`
```
2. Your `getContainerOfType` function actually uses `null`. We don't use `null` usually in Langium, but `undefined` instead. They have different semantic meanings (usually), and also behave differently in equality checks or `typeof` checks. See also [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/null#difference_between_null_and_undefined).
3. I believe this relatively simplistic way of finding previous variables stops working once you start working on multiple block levels and can become progressively complicated when you try to deal with those. Going at this problem from another perspective might be better: Instead of finding all elements that precede the current variable and see whether it contains the reference (which is also relatively computationally expensive), we can use the CST to do the job. Something like this should suffice:
```ts
export function isDefinedBefore(from: AstNode, to: AstNode): boolean {
  const fromDocument = getDocument(from);
  const toDocument = getDocument(to);
  // In case a variable from another document is referenced (might be illegal?)
  if (fromDocument !== toDocument) {
     return false;
  }
  if (!from.$cstNode || !to.$cstNode) {
    return false;
  }
  const fromPosition = from.$cstNode.offset;
  const toPosition = to.$cstNode.offset;
  return fromPosition < toPosition;
}
```
You can call that function using your expression in which the reference appears as `from` and the referenced variable as `to`.

**Reply by snarkipus** [reply]
Thanks @msujew 

Update: Unit Testing now works as expected - test code at the bottom 

> 1. ... which is typed a bit better than yours ...

Ha ... why yes, it is _just_ a bit better. ;)

> 2 ... We don't use `null` ... They have different semantic meanings (usually)

Yeah, I was actually looking at lodash for equality checking ... but I was pretty far down the rabbit hole when I got to that point.

> 3 ... Instead of finding all elements that precede the current variable and see whether it contains the reference (which is also relatively computationally expensive), we can use the CST to do the job. Something like this should suffice:

I keep forgetting about all of the tooling you guys have baked into the framework. I should probably try to ask questions before I charge off to reinvent the wheel.

I do have a couple questions about using this CST-based approach. I think I might have gone down an AST path when determining the `from` and `to` arguments when I should have gone down the CST path. Regardless, I'm stuck. (see testing code below)

```ts
describe('Model Utility Class: isDefinedBefore(from,to)', () => {
    it('should return true for valid reference', async () => {
        const text =`
        var  i = 0  // (0)
		eval i + 10 // (1)
        `; 

        const parse = await helper(text);
        const elements = parse.parseResult.value.elements;
		const from =  elements[0];
		const to   = (elements[1].expression as Plus).left as VariableRef;
		expect(isDefinedBefore(from, to)).toBe(true);
	});

	it('should return false for invalid forward reference', async () => {
        const text =`
		eval i + 10 // (0)
        var  i = 0  // (1)
		`; 

        const parse = await helper(text);
        const elements = parse.parseResult.value.elements;
		const from =  elements[1];
		const to   = (elements[0].expression as Plus).left as VariableRef;
		expect(isDefinedBefore(from, to)).toBe(false);
	});
});
```

**Reply by snarkipus** [reply]
@msujew  

Maybe trying to test it directly isn't the smartest approach out of the gate .. this 'mostly' works as part of the validation service. I'll have to think on the parts that don't work (see screenshot) ... maybe you've got some suggestions. Probably needs some sort of crappy recursion and state information ...

<img width="957" alt="image" src="https://user-images.githubusercontent.com/3688989/184520252-1a9dae20-6007-4218-8c29-f88deea9b589.png">

```ts
export namespace IssueCodes {
    export const ForwardReference = 'forward-reference';
}

/**
 * Implementation of custom validations.
 */
export class ExpressionsValidator {
    checkForwardReference(varRef: VariableRef, accept: ValidationAcceptor): void {
        const variable = varRef.variable.ref;
        if (!variable)  {
            accept(
                'error',
                'variable forward reference not allowed.',
                {
                    node: varRef,
                    property: 'variable',
                    code: IssueCodes.ForwardReference
                }
            );
        } else if (!isDefinedBefore(variable, varRef)) {
            accept(
                'error',
                'variable forward reference not allowed.',
                {
                    node: varRef,
                    property: 'variable',
                    code: IssueCodes.ForwardReference
                }
            );
        }
    }
}
```

**Reply by msujew** [reply]
So the `var p = p + 1` can be relatively easily addressed by looking for the variable declaration inside of the validation:

```ts
export class ExpressionsValidator {
    checkForwardReference(varRef: VariableRef, accept: ValidationAcceptor): void {
        let parent: AstNode | undefined = getContainerOfType(varRef, isVariable);
        
        if (!parent) {
            parent = varRef;
        }

        const variable = varRef.variable.ref;
        if (variable) {
            if (parent === variable) {
                accept('error', 'Variable cannot be used in own declaration', { node: varRef, property: 'variable' });
            } else if (isDefinedBefore(parent, variable)) {
                accept(
                    'error',
                    'variable forward reference not allowed.',
                    {
                        node: varRef,
                        property: 'variable',
                        code: IssueCodes.ForwardReference
                    }
                );
            }
        }
    }
}
```

I've removed the validation for `!variable`, since you get an linking error from Langium for free anyway.

Regarding the `var m = n + 1` part. This should be a legal operation, except if you don't want to allow cyclic assignments at all, something like:

```
var m = 1
var n = m + 1
m = n + 1
```

That's valid in any general purpose programming language. If you want to prevent that, you need to perform a depth first search on the AST graph. We do something similar while trying to figure out if a grammar contains unused rules:
https://github.com/langium/langium/blob/c88a58f9770137550ea4e6e4694db7497c54b6f5/packages/langium/src/grammar/langium-grammar-validator.ts#L532-L561

**Reply by snarkipus** [reply]
> ... easily addressed by looking for the variable declaration inside of the validation:

That's really clean ... well done. Hopefully at some point I'll stop fighting against the typing and start taking advantage of it. 

> ... except if you don't want to allow cyclic assignments at all ...

You would think having just completed a cycle detection validator for the previous chapter in this book ([entity hierarchy validation](https://github.com/snarkipus/langium-entities/blob/15c5c15359e5a94f0176184e41136c2de734b723/src/language-server/entities-validator.ts#L37-L59)) that a separate cycle validator would have been completely obvious. Regardless, thanks for linking the DFS example - that will very likely come in handy at some point.

Two completely unrelated and somewhat silly questions:
1) How do you get the github markdown to show a preview of the code you link to? I know how to get a permalink to code sections, but that just gives me a hyperlink.
2) How do I build my langium project against `langium@next` as to avoid tripping over bugs/issue already addressed? Or is this very much frowned upon?

Thanks as usual @msujew.

**Reply by msujew** [reply]
> How do you get the github markdown to show a preview of the code you link to? I know how to get a permalink to code sections, but that just gives me a hyperlink.

GitHub doesn't really explain it, but permalinking code from the __current repository__ (i.e. `langium/langium`) makes it appear as a preview. Permalinking code from other repos just shows the link. It's kind of a hidden feature, and took me some time to figure it out as well.

> How do I build my langium project against langium@next as to avoid tripping over bugs/issue already addressed? Or is this very much frowned upon?

You can just look at the versions available at [npmjs](https://www.npmjs.com/package/langium) and pick the newest (currently `0.5.0-next.f2b3802`). Since we don't regularly update our next versions, I would recommend something different, which I personally like to use a lot:

Instead of putting a version available from npm into your package json, you can also use a [local path](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#local-paths). So I clone langium and another project into the same directory and link langium like this:

```ts
"dependencies": {
  "langium": "file://../langium/packages/langium"
},
"devDependencies": {
  "langium-cli": "file://langium/packages/langium-cli"
}
```

That way all changes that I perform in Langium are also automatically available in the language I'm currently working on. Really helpful when working on a framework feature, which doesn't have a good default implementation or cannot be easily implemented using the existing example languages. There's also a [short section in our contribution guide](https://github.com/langium/langium/blob/main/CONTRIBUTING.md#altered-langium-for-own-language-projects) on this topic.

**Reply by snarkipus** [reply]
@msujew 

Hmmm ... well, that didn't work at all. In fact, I can't even get the hello-world project to build using a self-built langium. If this is a bug, let me know and I'll move it over. I suspect operator error of some kind though.
1) Clone latest langium and `cd langium`
2) `npm install`
3) `npm run dev-build`
4) `npm run test` (just to be sure)
```term
Test Suites: 37 passed, 37 total
Tests:       388 passed, 388 total
Snapshots:   0 total
Time:        17.746 s
Ran all test suites.
```
5) change to different directory, then `yo langium` with defaults
6) `cd hello-world`
7) `npm link -S langium`
8) `npm link -D langium-cli`
9) `npm run langium:generate` (successful)
10) `npm run build` <---- angry langium build

#### Error:
```
❯ npm run build

> hello-world@0.0.1 build
> tsc -b tsconfig.json

src/language-server/main.ts:10:47 - error TS2322: Type 'import("/home/snarkipus/Projects/langium/fresh/hello-world/node_modules/vscode-languageserver/lib/common/server")._Connection<import("/home/snarkipus/Projects/langium/fresh/hello-world/node_modules/vscode-languageserver/lib/common/server")._, import("/home/snarkipus/Projects/langium/fresh/hello-world/node_modules/vs...' is not assignable to type 'import("/home/snarkipus/Projects/langium/fresh/langium/node_modules/vscode-languageserver/lib/common/server")._Connection<import("/home/snarkipus/Projects/langium/fresh/langium/node_modules/vscode-languageserver/lib/common/server")._, import("/home/snarkipus/Projects/langium/fresh/langium/node_modules/vscode-languag...'.
  Types of property 'onRequest' are incompatible.
    Type '{ <R, PR, E, RO>(type: import("/home/snarkipus/Projects/langium/fresh/hello-world/node_modules/vscode-languageserver-protocol/lib/common/messages").ProtocolRequestType0<R, PR, E, RO>, handler: import("/home/snarkipus/Projects/langium/fresh/hello-world/node_modules/vscode-jsonrpc/lib/common/connection").RequestHandle...' is not assignable to type '{ <R, PR, E, RO>(type: import("/home/snarkipus/Projects/langium/fresh/langium/node_modules/vscode-languageserver-protocol/lib/common/messages").ProtocolRequestType0<R, PR, E, RO>, handler: import("/home/snarkipus/Projects/langium/fresh/langium/node_modules/vscode-jsonrpc/lib/common/connection").RequestHandler0<...>)...'.
      Types of parameters 'type' and 'type' are incompatible.
        Type 'import("/home/snarkipus/Projects/langium/fresh/langium/node_modules/vscode-languageserver-protocol/lib/common/messages").ProtocolRequestType0<any, any, any, any>' is not assignable to type 'import("/home/snarkipus/Projects/langium/fresh/hello-world/node_modules/vscode-languageserver-protocol/lib/common/messages").ProtocolRequestType0<any, any, any, any>'.
          Types of property 'parameterStructures' are incompatible.
            Type 'import("/home/snarkipus/Projects/langium/fresh/langium/node_modules/vscode-jsonrpc/lib/common/messages").ParameterStructures' is not assignable to type 'import("/home/snarkipus/Projects/langium/fresh/hello-world/node_modules/vscode-jsonrpc/lib/common/messages").ParameterStructures'.
              Types have separate declarations of a private property 'kind'.

10 const { shared } = createHelloWorldServices({ connection, ...NodeFileSystem });
                                                 ~~~~~~~~~~

  ../langium/packages/langium/lib/default-module.d.ts:28:5
    28     connection?: Connection;
           ~~~~~~~~~~
    The expected type comes from property 'connection' which is declared here on type 'DefaultSharedModuleContext'


Found 1 error.
```
npm config:
```term
❯ npm list
hello-world@0.0.1 /home/snarkipus/Projects/langium/fresh/hello-world
├── @types/node@14.18.24
├── @types/vscode@1.70.0
├── @typescript-eslint/eslint-plugin@5.33.1
├── @typescript-eslint/parser@5.33.1
├── chalk@4.1.2
├── chevrotain@9.1.0
├── commander@8.3.0
├── eslint@8.22.0
├── langium-cli@0.4.0 -> ./../langium/packages/langium-cli
├── langium@0.4.0 -> ./../langium/packages/langium
├── typescript@4.7.4
├── vscode-languageclient@8.0.2
├── vscode-languageserver@8.0.2
└── vscode-uri@3.0.3

❯ npm list -g langium
/home/snarkipus/.nvm/versions/node/v16.14.0/lib
├─┬ langium-cli@0.4.0 -> ./../../../../../Projects/langium/fresh/langium/packages/langium-cli
│ └── langium@0.4.0 -> ./../../../../../Projects/langium/fresh/langium/packages/langium
└─┬ langium@0.4.0 -> ./../../../../../Projects/langium/fresh/langium/packages/langium
  └─┬ langium-cli@0.4.0 -> ./../../../../../Projects/langium/fresh/langium/packages/langium-cli
    └── langium@0.4.0 deduped -> ./../../../../../Projects/langium/fresh/langium/packages/langium
```

**Reply by msujew** [reply]
There has been a breaking change in order to support web based Langium server. You can take a look at our [examples](https://github.com/langium/langium/blob/fd398f94d93399ab531316573cb4be22953834a2/examples/domainmodel/src/language-server/main.ts#L13) on how to resolve it. Basically, what kind of object we pass into our internal DI functions changed a bit.

**Reply by snarkipus** [reply]
Interesting, the resolution for the breaking change must not have been incorporated (yet) into the yeoman template for the `hello-world` default project. It works as described using the example projects linked against local langium packages.

**Reply by snarkipus** [reply]
@msujew I'm not smart enough to figure it out I suppose - the hello-world example works just fine if I run it inside the main langium project (i.e. in the examples directory). Trying to do the same thing outside of it results in the error I showed earlier. Not a huge deal, but I do wish I understood why it's dying. I can't see that it's related to the change you linked necessarily.

**Reply by msujew** [reply]
Actually, I didn't properly read your message (we currently have a week-long team event at TypeFox, little time to actually work). The issue is that we updated the `vscode-languageserver` library to `8.0.1`, which isn't compatible with the old version (7.0). I created a small patch for you to apply to your `langium-expressions` repo: [patch.zip](https://github.com/langium/langium/files/9371970/mypatch.zip)

**Reply by snarkipus** [reply]
No worries @msujew  - thank you very much for the patch, I'll give it a shot. In the meantime, since it's a relatively small example language, I just created a dev branch and recreated the project inside the workspace ([here](https://github.com/snarkipus/langium-dev/tree/local-work/examples/expressions)). That will have to do for now until I can figure out a more elegant solution.

Ideally, I'd be able to keep them separate since this approach breaks some of the dev tooling (i.e. Wallaby.js).
